### PR TITLE
[BE] Enforce sign-compare

### PR DESCRIPTION
**Pull request by malfet**
Number of OSS PR were reverted, because new signed-unsigned comparison warnings, which are treated as errors in some internal builds.
Not sure how those selective rules are applied, but this PR removes `-Wno-sign-compare` from PyTorch codebase.

The only tricky part in this PR, as making sure that non-ASCII character detection works for both signed and unsigned chars  here:
https://github.com/pytorch/pytorch/blob/6e3d51b08ac108b27d892ab1be85eeb593ec1f0c/torch/csrc/jit/serialization/python_print.cpp#L926

Exclude several files from sign-compare if flash attention is used, due to the violation in cutlass, to be fixed by https://github.com/NVIDIA/cutlass/pull/869
Do not try to fix sign compare violations in caffe2 codebase

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @EikanWang

### CODE REVIEW

File: caffe2/perfkernels/fused_nbit_rowwise_conversion.cc [hunk 1]
```diff
@@ -59,7 +58,7 @@ void Fused8BitRowwiseQuantizedToFloat__base(
     float* output_row = output + row * output_columns;
 
     // NOLINTNEXTLINE(clang-diagnostic-sign-compare)
-    for (std::size_t col = 0; col < output_columns; ++col) {
+    for (std::size_t col = 0; col < static_cast<std::size_t>(output_columns); ++col) {
```
Comment (albanD) [inline caffe2/perfkernels/fused_nbit_rowwise_conversion.cc:61]: Comment above can be removed.

**Review by albanD — COMMENTED** [review]
Current change look pretty good! :)

File: torch/csrc/distributed/autograd/rpc_messages/rpc_with_profiling_resp.cpp [hunk 2]
```diff
@@ -125,7 +125,7 @@ std::unique_ptr<RpcWithProfilingResp> RpcWithProfilingResp::fromMessage(
            kProfileEventsStartIdx,
            kProfileEventsStartIdx + profiledEventsSize)) {
     // NOLINTNEXTLINE(clang-diagnostic-sign-compare)
-    TORCH_CHECK(i < tupleElements.size());
+    TORCH_CHECK(static_cast<size_t>(i) < tupleElements.size());
```
Comment (albanD) [inline torch/csrc/distributed/autograd/rpc_messages/rpc_with_profiling_resp.cpp:128]: Comment

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 3]
```diff
@@ -504,7 +502,7 @@ class PrioritizeLoad : public IRMutator {
         // also check indices
         bool same = true;
         // NOLINTNEXTLINE(clang-diagnostic-sign-compare)
-        for (int i = 0; i < v->indices().size(); ++i) {
+        for (const auto i : c10::irange(v->indices().size())) {
```
Comment (albanD) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:505]: comment

**Review by albanD — APPROVED** [review]
Thanks for taking care of that!

**Comment by xwang233** [comment]
cc @jjsjann123

**Pull request closed without merging**